notebooks/ch17/classification_multiclass.md

```diff
@@ -0,0 +1,312 @@
+- Iris-virginica (`2`)
+
+There are four features available in the dataset:
+- Sepal length (cm)
+- Sepal width (cm)
+- Petal length (cm)
+- Petal width (cm)
+
+![](petal_sepal.png)
+
+We will create a multiclass classifier that predicts the type of Iris plant based on the four features above. First, we read in the data:
+
+
+```python
+iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data',
+                  header=None, names=['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species'])
+
+iris.head()
+```
+
+
+
+
+<div>
+<style scoped>
+    .dataframe tbody tr th:only-of-type {
+        vertical-align: middle;
+    }
+
+    .dataframe tbody tr th {
+        vertical-align: top;
+    }
+
+    .dataframe thead th {
+        text-align: right;
+    }
+</style>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>sepal_length</th>
+      <th>sepal_width</th>
+      <th>petal_length</th>
+      <th>petal_width</th>
+      <th>species</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>5.1</td>
+      <td>3.5</td>
+      <td>1.4</td>
+      <td>0.2</td>
+      <td>Iris-setosa</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>4.9</td>
+      <td>3.0</td>
+      <td>1.4</td>
+      <td>0.2</td>
+      <td>Iris-setosa</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>4.7</td>
+      <td>3.2</td>
+      <td>1.3</td>
+      <td>0.2</td>
+      <td>Iris-setosa</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>4.6</td>
+      <td>3.1</td>
+      <td>1.5</td>
+      <td>0.2</td>
+      <td>Iris-setosa</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>5.0</td>
+      <td>3.6</td>
+      <td>1.4</td>
+      <td>0.2</td>
+      <td>Iris-setosa</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+
+```python
+X, y = iris.drop('species', axis=1), iris['species']
+```
+
+
+```python
+# Alternatively, read in from sklearn.datasets
+# from sklearn.datasets import load_iris
+
+# X, y = load_iris(return_X_y=True)
+```
+
+
+```python
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, random_state=42)
+```
+
+After dividing the dataset into train and test splits, we fit a multiclass classifier to our training data. By default, scikit-learn's `LogisticRegression` sets `multi_class='ovr'`:
+
+
+```python
+lr = LogisticRegression(random_state=42)
+lr.fit(X_train, y_train)
+```
+
+
+
+
+    LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
+              intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
+              penalty='l2', random_state=42, solver='liblinear', tol=0.0001,
+              verbose=0, warm_start=False)
+
+
+
+We predict on the test data, and use a confusion matrix to evaluate the results:
+
+
+```python
+# to-do: add x-axis, y-axis labels to confusion matrix
+y_pred = lr.predict(X_test)
+confusion_matrix(y_test, y_pred)
```

Comment:
Can you use https://seaborn.pydata.org/generated/seaborn.heatmap.html to display a heatmap of the matrix with the axis labels and such? It'll be easier to interpret.

Reply:
The output didn't show up in the markdown file, but the confusion matrix looks like this:
![image](https://user-images.githubusercontent.com/25359351/42722112-452e26cc-86fb-11e8-938d-628bc35fe3a8.png)
So the seaborn heatmaps don't show up well:
![image](https://user-images.githubusercontent.com/25359351/42722118-6485f914-86fb-11e8-9c3a-60ddbf03b682.png)
I think keeping it as a table with numbers and axis labels might be easier to interpret. What do you think?

(I found this pretty funny, actually; I had to deliberately set the test size to 0.35 in order to get any misclassifications at all. The dataset doesn't pose much challenge to logistic regression, unfortunately.)

Reply:
Yeah I ran into a similar problem with the spam/ham dataset, so I increased the number of data points (which was possible since the email database is huge) and tried different model features until I got one that would be useful for instructional purposes. I don't think either of these are possible for the iris data (considering all features are given), so this is probably fine. Not sure how the `sns.heatmap` colors work but I think blue is a good color.